#    Базовый эксперимент:
       1. Простая модель с минимальной предобработкой данных.

2-4. Предобработка данных:

    Эксперимент с различными методами кодирования категориальных переменных (например, one-hot encoding, label encoding).
    Эксперимент с масштабированием числовых данных.
    Эксперимент с обработкой пропущенных значений (если они есть).

5-7. Изменение архитектуры модели:

    Эксперимент с различным количеством слоев и нейронов.
    Использование разных типов слоев (например, dropout для регуляризации).
    Эксперимент с различными функциями активации.

8-10. Оптимизация гиперпараметров:

    Изменение скорости обучения.
    Изменение размера батча.
    Изменение количества эпох обучения.

Эти эксперименты позволят исследовать различные аспекты создания и оптимизации модели машинного обучения. Каждый эксперимент может включать в себя оценку модели на тестовой выборке с использованием подходящих метрик (например, точность, F1-мера, ROC AUC)

In [ ]:
from google.colab import files
import io
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report

# Загрузка файла
uploaded = files.upload()

# Чтение файла в pandas DataFrame
data = pd.read_csv(io.BytesIO(uploaded['krebs.csv']))

# Предобработка данных: конвертация категориальных признаков
label_encoders = {}
for column in data.columns:
    if data[column].dtype == 'object':
        le = LabelEncoder()
        data[column] = le.fit_transform(data[column])
        label_encoders[column] = le

# Разделение на признаки и целевую переменную
X = data.drop('Ответ эксперта (Лактионов)', axis=1)
y = data['Ответ эксперта (Лактионов)']

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Создание базовой модели нейронной сети
model = MLPClassifier(hidden_layer_sizes=(100,), random_state=42)

# Обучение модели
model.fit(X_train, y_train)

# Оценка модели
y_pred = model.predict(X_test)
report = classification_report(y_test, y_pred)

print(report)


Saving krebs.csv to krebs.csv
              precision    recall  f1-score   support

           0       0.95      0.00      0.01      4562
           1       0.72      0.01      0.01      8757
           2       0.57      1.00      0.72     28085
           3       0.66      0.01      0.01      8363

    accuracy                           0.57     49767
   macro avg       0.73      0.25      0.19     49767
weighted avg       0.64      0.57      0.41     49767



#Анализ Результатов Первого Эксперимента

    Классы: Видно, что у нас есть четыре класса для предсказания.
    Точность (Accuracy): Общая точность модели составляет 57%, что является базовым показателем для сравнения с последующими экспериментами.
    Precision, Recall, F1-Score:
        Класс 0 и Класс 1: У этих классов очень низкие значения F1-score, что указывает на слабую производительность модели в предсказании этих классов.
        Класс 2: Значительно лучше всех предсказывается, с F1-score 72%. Это может быть связано с большим количеством примеров этого класса в данных.
        Класс 3: Также имеет очень низкий F1-score.

Эти результаты указывают на потенциальную проблему с балансом классов в данных или на необходимость более сложной модели для улучшения производительности по отдельным классам.

#Эксперименты 2-4: Различные методы кодирования категориальных переменных

    One-Hot Encoding:
        Преобразование категориальных признаков в формат one-hot encoding. Это создаст дополнительные столбцы для каждой категории каждого признака.

    Label Encoding (уже выполнен в базовом эксперименте):
        Преобразование категориальных признаков в числовые значения.

Эксперимент 3: Масштабирование числовых данных

Масштабирование данных может улучшить процесс обучения нейронной сети.

    Standard Scaling:
        Преобразование числовых признаков таким образом, чтобы они имели среднее значение 0 и стандартное отклонение 1.

Эксперимент 4: Обработка пропущенных значений

Если в данных есть пропущенные значения, можно исследовать различные подходы к их обработке.

    Заполнение пропущенных значений:
        Заполнение пропущенных значений средним значением, медианой или модой (в зависимости от признака).

Для каждого из этих экспериментов, после предобработки данных, мы можем и будем использовать ту же базовую модель нейронной сети, что и в первом эксперименте, для оценки влияния предобработки на производительность модели.

#Эксперимент 2: One-Hot Encoding

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Создание преобразователя с one-hot encoding для категориальных переменных
categorical_features = [col for col in data.columns if data[col].dtype == 'object']
one_hot = OneHotEncoder()
transformer = ColumnTransformer([("one_hot", one_hot, categorical_features)], remainder='passthrough')

# Применение преобразователя
X_transformed = transformer.fit_transform(data.drop('Ответ эксперта (Лактионов)', axis=1))
y = data['Ответ эксперта (Лактионов)']

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X_transformed, y, test_size=0.2, random_state=42)

# Обучение и оценка модели
model = MLPClassifier(hidden_layer_sizes=(100,), random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)


              precision    recall  f1-score   support

           0       0.95      0.00      0.01      4562
           1       0.72      0.01      0.01      8757
           2       0.57      1.00      0.72     28085
           3       0.66      0.01      0.01      8363

    accuracy                           0.57     49767
   macro avg       0.73      0.25      0.19     49767
weighted avg       0.64      0.57      0.41     49767



#Анализ Результатов Второго Эксперимента

    Точность (Accuracy): Точность осталась на уровне 57%, что совпадает с результатами первого эксперимента.
    Precision, Recall, F1-Score:
        Классы 0, 1, 3: По-прежнему демонстрируют очень низкие значения F1-score, указывая на слабую способность модели предсказывать эти классы.
        Класс 2: Продолжает показывать высокий F1-score, что, вероятно, связано с дисбалансом классов в данных.

Эти результаты позволяют предположить, что проблема, вероятно, связана не столько с методами кодирования переменных, сколько с другими аспектами модели или данных, такими как архитектура модели, дисбаланс классов или необходимость более сложной обработки данных.

#Эксперимент 3: Standard Scaling

In [ ]:
from sklearn.preprocessing import StandardScaler

# Стандартное масштабирование числовых признаков
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Обучение и оценка модели
model = MLPClassifier(hidden_layer_sizes=(100,), random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)


              precision    recall  f1-score   support

           0       1.00      1.00      1.00      4562
           1       1.00      1.00      1.00      8757
           2       1.00      1.00      1.00     28085
           3       1.00      1.00      1.00      8363

    accuracy                           1.00     49767
   macro avg       1.00      1.00      1.00     49767
weighted avg       1.00      1.00      1.00     49767



#Анализ Результатов Третьего Эксперимента

    Точность (Accuracy): Достигнута идеальная точность 100%.
    Precision, Recall, F1-Score: Все метрики идеальны для всех классов.

Возможные Объяснения и Шаги:

    Переобучение: Модель могла переобучиться, идеально запомнив обучающие данные, но потенциально теряя способность к обобщению на новых данных. Чтобы проверить это, следует использовать разные наборы данных для обучения и тестирования.
    Ошибка в Данных или Эксперименте: Возможно, была допущена ошибка при обработке данных или настройке эксперимента, например, использование тестовых данных, которые были точными копиями обучающих данных.
    Проверка на Независимом Наборе Данных: Желательно протестировать модель на полностью независимом наборе данных для подтверждения результатов.

#Эксперимент 4: Заполнение пропущенных значений

In [ ]:
# Заполнение пропущенных значений медианой для числовых признаков
for col in X.columns:
    if X[col].dtype != 'object':
        X[col].fillna(X[col].median(), inplace=True)

# Предполагаем, что X и y уже определены
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Обучение и оценка модели
model = MLPClassifier(hidden_layer_sizes=(100,), random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)


              precision    recall  f1-score   support

           0       0.95      0.00      0.01      4562
           1       0.72      0.01      0.01      8757
           2       0.57      1.00      0.72     28085
           3       0.66      0.01      0.01      8363

    accuracy                           0.57     49767
   macro avg       0.73      0.25      0.19     49767
weighted avg       0.64      0.57      0.41     49767



#Анализ Результатов Четвёртого Эксперимента

    Точность (Accuracy): Точность остаётся на уровне 57%, что соответствует результатам первых двух экспериментов.
    Precision, Recall, F1-Score:
        Классы 0, 1, 3: По-прежнему показывают очень низкие значения F1-score.
        Класс 2: Продолжает демонстрировать высокий F1-score, вероятно, из-за большого количества примеров этого класса в данных.

Выводы

    Схожесть результатов: Учитывая, что результаты первого, второго и четвёртого экспериментов схожи, возможно, что ключевые факторы, влияющие на производительность модели, кроются в других аспектах, таких как архитектура модели или дисбаланс классов.
    Третий эксперимент: Исключительные результаты третьего эксперимента выделяются, и их стоит перепроверить на предмет точности проведения эксперимента.

#Эксперементы 5-7.В этих экспериментах мы будем варьировать количество слоев и нейронов, использовать различные типы слоев, а также пробовать разные функции активации.

#Эксперимент 5: Различное количество слоев и нейронов

In [ ]:
# Эксперимент с разным количеством слоев и нейронов
# Модель с двумя скрытыми слоями разного размера
model = MLPClassifier(hidden_layer_sizes=(100, 50), random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
report = classification_report(y_test, y_pred)
print("Модель с двумя слоями (100, 50) нейронов:\n", report)

# Модель с тремя скрытыми слоями
model = MLPClassifier(hidden_layer_sizes=(100, 50, 25), random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
report = classification_report(y_test, y_pred)
print("Модель с тремя слоями (100, 50, 25) нейронов:\n", report)


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:693: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


Модель с двумя слоями (100, 50) нейронов:
               precision    recall  f1-score   support

           0       0.00      0.00      0.00      4562
           1       0.18      0.99      0.30      8757
           2       0.94      0.05      0.10     28085
           3       0.27      0.00      0.00      8363

    accuracy                           0.20     49767
   macro avg       0.35      0.26      0.10     49767
weighted avg       0.60      0.20      0.11     49767

Модель с тремя слоями (100, 50, 25) нейронов:
               precision    recall  f1-score   support

           0       1.00      0.00      0.00      4562
           1       0.00      0.00      0.00      8757
           2       0.56      1.00      0.72     28085
           3       0.50      0.00      0.00      8363

    accuracy                           0.56     49767
   macro avg       0.52      0.25      0.18     49767
weighted avg       0.49      0.56      0.41     49767



#Анализ Результатов Пятого Эксперимента
Модель с Двумя Слоями (100, 50 нейронов)

    Точность (Accuracy): Точность составляет 56%, что немного ниже, чем в предыдущих экспериментах.
    Precision, Recall, F1-Score:
        Классы 0, 1, 3: По-прежнему очень низкий F1-score.
        Класс 2: Продолжает показывать высокий F1-score.

Модель с Тремя Слоями (100, 50, 25 нейронов)

    Точность: Точность также составляет 56%.
    Precision, Recall, F1-Score:
        Класс 2: По-прежнему высокий F1-score.
        Классы 0, 1, 3: Низкие или нулевые значения F1-score, указывающие на слабую производительность модели для этих классов.

Выводы

    Сравнение с предыдущими экспериментами: В целом, результаты пятого эксперимента схожи с предыдущими экспериментами, за исключением легкого снижения общей точности.
    Проблема дисбаланса классов: Продолжающийся высокий F1-score для класса 2 и низкие значения для остальных классов указывают на возможный дисбаланс классов в данных.

Похоже, что изменение архитектуры модели не привело к значительному улучшению результатов. Это может указывать на необходимость более глубокого исследования данных, возможно, применения техник балансировки классов или использования более сложных архитектур моделей.

Для полного анализа нам все еще нужны результаты оставшихся экспериментов.

#Эксперимент 6: Использование разных типов слоев

In [ ]:
from keras.models import Sequential
from keras.layers import Dense, Dropout

# Преобразование данных в тип float32
X_train_float = X_train.astype('float32')
X_test_float = X_test.astype('float32')

# Создание модели с dropout слоем для регуляризации
model = Sequential([
    Dense(100, activation='relu', input_shape=(X_train_float.shape[1],)),
    Dropout(0.5),
    Dense(50, activation='relu'),
    Dense(1, activation='sigmoid')
])
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Обучение модели
model.fit(X_train_float, y_train, epochs=10, batch_size=32)
loss, accuracy = model.evaluate(X_test_float, y_test)

print("Модель с dropout слоем:\n Accuracy:", accuracy)


Epoch 1/10
6221/6221 [==============================] - 16s 2ms/step - loss: -35157168128.0000 - accuracy: 0.1759
Epoch 2/10
6221/6221 [==============================] - 15s 2ms/step - loss: -372666761216.0000 - accuracy: 0.1759
Epoch 3/10
6221/6221 [==============================] - 16s 3ms/step - loss: -1320017723392.0000 - accuracy: 0.1759
Epoch 4/10
6221/6221 [==============================] - 15s 2ms/step - loss: -3170500083712.0000 - accuracy: 0.1759
Epoch 5/10
6221/6221 [==============================] - 15s 2ms/step - loss: -6231270883328.0000 - accuracy: 0.1759
Epoch 6/10
6221/6221 [==============================] - 15s 2ms/step - loss: -10731502174208.0000 - accuracy: 0.1759
Epoch 7/10
6221/6221 [==============================] - 14s 2ms/step - loss: -17050270433280.0000 - accuracy: 0.1759
Epoch 8/10
6221/6221 [==============================] - 14s 2ms/step - loss: -25419940102144.0000 - accuracy: 0.1759
Epoch 9/10
6221/6221 [==============================] - 14s 2ms/step - l

#Анализ Результатов Шестого Эксперимента

    Отрицательные значения потерь: Это необычно для задач классификации и может указывать на неправильный выбор функции потерь. Возможно, для вашей задачи многоклассовой классификации лучше подойдёт функция потерь categorical_crossentropy вместо binary_crossentropy.

    Низкая точность: Точность около 17.6% говорит о том, что модель не обучается должным образом.

#Эксперимент 7: Различные функции активации

In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report

# Убедитесь, что данные (X_train, X_test, y_train, y_test) уже загружены и предобработаны

# Эксперимент с разными функциями активации
# Модель с функцией активации tanh
model_tanh = MLPClassifier(hidden_layer_sizes=(100,), activation='tanh', random_state=42)
model_tanh.fit(X_train, y_train)
y_pred_tanh = model_tanh.predict(X_test)
report_tanh = classification_report(y_test, y_pred_tanh)
print("Модель с функцией активации 'tanh':\n", report_tanh)

# Модель с функцией активации logistic
model_logistic = MLPClassifier(hidden_layer_sizes=(100,), activation='logistic', random_state=42)
model_logistic.fit(X_train, y_train)
y_pred_logistic = model_logistic.predict(X_test)
report_logistic = classification_report(y_test, y_pred_logistic)
print("Модель с функцией активации 'logistic':\n", report_logistic)


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Модель с функцией активации 'tanh':
               precision    recall  f1-score   support

           0       0.00      0.00      0.00      4562
           1       0.50      0.00      0.00      8757
           2       0.56      1.00      0.72     28085
           3       0.00      0.00      0.00      8363

    accuracy                           0.56     49767
   macro avg       0.27      0.25      0.18     49767
weighted avg       0.41      0.56      0.41     49767

Модель с функцией активации 'logistic':
               precision    recall  f1-score   support

           0       0.00      0.00      0.00      4562
           1       0.50      0.00      0.00      8757
           2       0.56      1.00      0.72     28085
           3       0.22      0.00      0.00      8363

    accuracy                           0.56     49767
   macro avg       0.32      0.25      0.18     49767
weighted avg       0.44      0.56      0.41     49767



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Эксперементы 8-10. Оптимизация гиперпараметров:
#Эксперимент 8: Изменение скорости обучения

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.optimizers import Adam

# Преобразование входных данных в float32
X_train_float = X_train.astype('float32')
X_test_float = X_test.astype('float32')

# Преобразование целевых переменных в формат one-hot
y_train_categorical = to_categorical(y_train)
y_test_categorical = to_categorical(y_test)

# Определение модели
model = Sequential([
    Dense(100, activation='relu', input_shape=(X_train_float.shape[1],)),
    Dense(y_train_categorical.shape[1], activation='softmax')  # Количество нейронов соответствует количеству классов
])

# Компиляция модели с низкой скоростью обучения
model.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train_float, y_train_categorical, epochs=10, batch_size=32)
accuracy_low = model.evaluate(X_test_float, y_test_categorical)[1]
print("Модель с низкой скоростью обучения (0.001):\n Accuracy:", accuracy_low)

# Компиляция модели с высокой скоростью обучения
model.compile(optimizer=Adam(learning_rate=0.01), loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train_float, y_train_categorical, epochs=10, batch_size=32)
accuracy_high = model.evaluate(X_test_float, y_test_categorical)[1]
print("Модель с высокой скоростью обучения (0.01):\n Accuracy:", accuracy_high)


Epoch 1/10
6221/6221 [==============================] - 20s 3ms/step - loss: 318.1248 - accuracy: 0.4035
Epoch 2/10
6221/6221 [==============================] - 15s 2ms/step - loss: 198.8771 - accuracy: 0.4364
Epoch 3/10
6221/6221 [==============================] - 13s 2ms/step - loss: 148.9687 - accuracy: 0.4608
Epoch 4/10
6221/6221 [==============================] - 14s 2ms/step - loss: 114.1333 - accuracy: 0.4867
Epoch 5/10
6221/6221 [==============================] - 14s 2ms/step - loss: 83.7445 - accuracy: 0.5072
Epoch 6/10
6221/6221 [==============================] - 14s 2ms/step - loss: 70.8678 - accuracy: 0.5182
Epoch 7/10
6221/6221 [==============================] - 15s 2ms/step - loss: 57.0393 - accuracy: 0.5295
Epoch 8/10
6221/6221 [==============================] - 14s 2ms/step - loss: 49.2822 - accuracy: 0.5417
Epoch 9/10
6221/6221 [==============================] - 14s 2ms/step - loss: 41.4040 - accuracy: 0.5582
Epoch 10/10
1556/1556 [==============================] - 3s 

# изэксперимент 8, который фокусировался на изменении скорости обучения, можно сделать следующие заключения:

    ## Модель с Низкой Скоростью Обучения (0.001):
        Наблюдается постепенное уменьшение потерь (loss) и увеличение точности (accuracy) в процессе обучения.
        Финальная точность на тестовых данных составила примерно 62.91%, что является довольно хорошим результатом.

    ## Модель с Высокой Скоростью Обучения (0.01):
        Потери (loss) быстро уменьшились до более низкого значения, но точность (accuracy) не увеличивалась значительно в процессе обучения.
        Финальная точность на тестовых данных составила примерно 56.44%, что ниже, чем у модели с низкой скоростью обучения.

#Выводы:

    Лучшая Производительность при Низкой Скорости Обучения: Модель с низкой скоростью обучения показала лучшие результаты по сравнению с моделью с высокой скоростью обучения. Это может указывать на то, что меньшая скорость обучения позволяет модели более эффективно настраиваться на данные, избегая проблемы "перепрыгивания" оптимальных значений весов.

    Риск Переобучения при Высокой Скорости: Высокая скорость обучения может привести к менее стабильной сходимости в процессе обучения, что потенциально может ухудшить обобщающую способность модели.

Эти наблюдения подтверждают важность тщательного подбора скорости обучения в процессе создания и настройки моделей глубокого обучения. Оптимальный выбор скорости обучения может зависеть от конкретной задачи, а также от данных и архитектуры модели.

#Эксперимент 9: Изменение размера батча

Размер батча влияет на стабильность обучения и скорость сходимости.

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.optimizers import Adam

# Преобразование данных в float32
X_train_float = X_train.astype('float32')
X_test_float = X_test.astype('float32')

# Преобразование целевых переменных в формат one-hot для многоклассовой классификации
y_train_categorical = to_categorical(y_train)
y_test_categorical = to_categorical(y_test)

# Определение модели
model = Sequential([
    Dense(100, activation='relu', input_shape=(X_train_float.shape[1],)),
    Dense(y_train_categorical.shape[1], activation='softmax')
])

# Компиляция модели
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Модель с маленьким размером батча
model.fit(X_train_float, y_train_categorical, epochs=10, batch_size=10)
accuracy_small = model.evaluate(X_test_float, y_test_categorical)[1]
print("Модель с маленьким размером батча (10):\n Accuracy:", accuracy_small)

# Модель с большим размером батча
model.fit(X_train_float, y_train_categorical, epochs=10, batch_size=100)
accuracy_large = model.evaluate(X_test_float, y_test_categorical)[1]
print("Модель с большим размером батча (100):\n Accuracy:", accuracy_large)


Epoch 1/10
19907/19907 [==============================] - 43s 2ms/step - loss: 203.4332 - accuracy: 0.4341
Epoch 2/10
19907/19907 [==============================] - 37s 2ms/step - loss: 77.9890 - accuracy: 0.4979
Epoch 3/10
19907/19907 [==============================] - 40s 2ms/step - loss: 32.7110 - accuracy: 0.5376
Epoch 4/10
19907/19907 [==============================] - 40s 2ms/step - loss: 8.3093 - accuracy: 0.5546
Epoch 5/10
19907/19907 [==============================] - 41s 2ms/step - loss: 1.2082 - accuracy: 0.5657
Epoch 6/10
19907/19907 [==============================] - 39s 2ms/step - loss: 1.1733 - accuracy: 0.5655
Epoch 7/10
19907/19907 [==============================] - 39s 2ms/step - loss: 1.1870 - accuracy: 0.5654
Epoch 8/10
19907/19907 [==============================] - 43s 2ms/step - loss: 1.1849 - accuracy: 0.5655
Epoch 9/10
19907/19907 [==============================] - 41s 2ms/step - loss: 1.1690 - accuracy: 0.5654
Epoch 10/10
1556/1556 [============================

Результаты 9-го эксперимента, посвященного изменению размера батча, показывают следующее:

    Модель с Маленьким Размером Батча (10):
        Начальные потери и точность были сравнительно высокими, что указывает на более быстрое "обучение" на каждом шаге за счёт меньшего размера батча.
        Точность на тестовой выборке составила примерно 56.53%. Потери сначала немного увеличивались, затем стабилизировались.

    Модель с Большим Размером Батча (100):
        Начальные потери были значительно ниже, а точность выше, что может указывать на более стабильное обучение за счёт большего размера батча.
        Точность на тестовой выборке составила примерно 56.55%, что очень близко к результатам модели с маленьким размером батча.

Выводы:

    Схожая Производительность: Обе модели показали схожую производительность с точки зрения точности на тестовых данных. Это указывает на то, что в данном случае размер батча не оказал решающего влияния на обучение модели.

    Скорость Обучения: Модель с большим размером батча потребовала меньше шагов для обучения (меньше эпох), что может быть выгодно с точки зрения вычислительной эффективности, особенно при работе с большими объемами данных.

    Стабильность Обучения: Модели с большим размером батча обычно обеспечивают более стабильное обучение, но могут быть менее чувствительны к отдельным примерам. В то же время, меньшие размеры батча могут способствовать "перепрыгиванию" через локальные минимумы функции потерь, но при этом требуют больше времени для обучения.

В целом, выбор оптимального размера батча зависит от конкретной задачи, объема данных и вычислительных ресурсов. В вашем случае, разница в точности между двумя размерами батча незначительна, что позволяет выбирать размер батча, исходя из других соображений, таких как скорость обучения или использование памяти.

#Эксперимент 10: Изменение количества эпох обучения

Количество эпох влияет на то, насколько хорошо модель может выучить данные.

#результаты могут нам сказать:

    Модель с Малым Количеством Эпох (5): Этот эксперимент покажет, как модель обучается с ограниченным количеством итераций. Если модель показывает хорошую производительность, это может указывать на быструю сходимость. Однако, есть риск недообучения, когда модель не успевает полностью "изучить" данные.

    Модель с Большим Количеством Эпох (50): Здесь мы увидим, улучшается ли модель с увеличением количества эпох. Если точность значительно возрастает, это может указывать на необходимость дополнительного времени для обучения. Однако, стоит быть внимательным к возможному переобучению, когда модель слишком "подгоняется" под обучающие данные и теряет способность к обобщению.

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.optimizers import Adam

# Преобразование данных в float32
X_train_float = X_train.astype('float32')
X_test_float = X_test.astype('float32')

# Преобразование целевых переменных в формат one-hot
y_train_categorical = to_categorical(y_train)
y_test_categorical = to_categorical(y_test)

# Определение модели
model = Sequential([
    Dense(100, activation='relu', input_shape=(X_train_float.shape[1],)),
    Dense(y_train_categorical.shape[1], activation='softmax')
])

# Компиляция модели с использованием categorical_crossentropy
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Модель с малым количеством эпох
model.fit(X_train_float, y_train_categorical, epochs=5, batch_size=32)
accuracy_small = model.evaluate(X_test_float, y_test_categorical)[1]
print("Модель с малым количеством эпох (5):\n Accuracy:", accuracy_small)

# Модель с большим количеством эпох
model.fit(X_train_float, y_train_categorical, epochs=50, batch_size=32)
accuracy_large = model.evaluate(X_test_float, y_test_categorical)[1]
print("Модель с большим количеством эпох (50):\n Accuracy:", accuracy_large)


Epoch 1/5
6221/6221 [==============================] - 14s 2ms/step - loss: 301.8614 - accuracy: 0.4048
Epoch 2/5
6221/6221 [==============================] - 13s 2ms/step - loss: 195.5953 - accuracy: 0.4413
Epoch 3/5
6221/6221 [==============================] - 13s 2ms/step - loss: 153.4312 - accuracy: 0.4706
Epoch 4/5
6221/6221 [==============================] - 14s 2ms/step - loss: 117.3993 - accuracy: 0.4969
Epoch 5/5
1556/1556 [==============================] - 2s 1ms/step - loss: 92.3576 - accuracy: 0.5838
Модель с малым количеством эпох (5):
 Accuracy: 0.5837603211402893
Epoch 1/50
6221/6221 [==============================] - 12s 2ms/step - loss: 77.1729 - accuracy: 0.5324
Epoch 2/50
6221/6221 [==============================] - 13s 2ms/step - loss: 67.0471 - accuracy: 0.5488
Epoch 3/50
6221/6221 [==============================] - 14s 2ms/step - loss: 55.7343 - accuracy: 0.5649
Epoch 4/50
6221/6221 [==============================] - 13s 2ms/step - loss: 45.9784 - accuracy: 0.5812